### PR TITLE
Implement tesh-timeout and tesh-long-running

### LINKED ISSUE #24: Support setting TIMEOUT
Our default timeout is 30 secs. This should be configurable.

### PR DESCRIPTION
**Pull request by domenkozar**
Fixes #24

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,7 +112,9 @@ You can set a few other optional directives in the header line:
 - `tesh-setup`: a filename of a script to run before running the commands in the code block,
 - `tesh-ps1`: allow an additional PS1 prompt besides the default `$`,
 - `tesh-platform`: specify on which platforms this session block should be tested (`linux`, `darwin`, `windows`),
-- `tesh-fixture`: a filename to save the current snippet.
+- `tesh-fixture`: a filename to save the current snippet,
+- `tesh-timeout`: number of second before a command timeouts (defaults to 30s),
```
Comment (zupo) [inline README.md:116]:
```suggestion
- `tesh-timeout`: number of seconds before a command timeouts (defaults to 30s),
```

File: README.md [hunk 2]
```diff
@@ -112,7 +112,9 @@ You can set a few other optional directives in the header line:
...
+- `tesh-timeout-expected`: if set to yes, the commands are expected to timeout (useful for showing long running processes).
```
Comment (zupo) [inline README.md:117]:
```suggestion
- `tesh-timeout-expected`: if set to `yes` or `true`, the commands are expected to timeout (useful for showing long running processes).
```

File: src/tesh/extract.py [hunk 3]
```diff
@@ -125,11 +127,16 @@ def extract(
...
-                    if setup:
+                    if setup := directives.get("setup", None):
                         # TODO: should we also check if setup would be overridden?
                         session.setup = setup
 
+                    if timeout := directives.get("timeout"):
+                        session.timeout = int(timeout)
+
+                    if timeout_expected := directives.get("timeout-expect"):
+                        session.timeout_expected = timeout_expected == "yes"
```
Comment (zupo) [inline src/tesh/extract.py:138]:
```suggestion
                        import distutils
                        session.timeout_expected = bool(distutils.util.strtobool(timeout_expected))
```

File: src/tesh/tests/fixtures/timeout.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Drop into an interactive shell on timeout
 
-```shell-session tesh-session="foo"
+```shell-session tesh-session="foo" tesh-timeout="1"
 $ sleep 35
```
Comment (zupo) [inline src/tesh/tests/fixtures/timeout.md:4]:
```suggestion
$ sleep 3
```

**Comment by zupo** [comment]
A few minor suggestions posted as inline comments. Gimme +1 / -1 on them, and I'll polish them in, update tests, etc.

**Comment by zupo** [comment]
I don't fully understand the usecase for `tesh-timeout-expected`. Couldn't I just set a very long timeout and effectively have the same outcome? Why do I need an extra directive?

**Comment by domenkozar** [comment]
> I don't fully understand the usecase for `tesh-timeout-expected`. Couldn't I just set a very long timeout and effectively have the same outcome? Why do I need an extra directive?

If you set a long timeout, the command will eventually timeout and fail.

If you run a long-running process that never exists (for example in my use case, `devenv up`), you want to kill the program eventually.

I need to look if we can inspect the output after timeout happens and make sure that only the final command can timeout.

**Comment by zupo** [comment]
Maybe `tesh-long-running` instead of `tesh-timeout-expected`?

File: src/tesh/extract.py [hunk 5]
```diff
@@ -125,11 +127,17 @@ def extract(
...
-                    setup = directives.get("setup", None)
-                    if setup:
+                    if setup := directives.get("setup", None):
                         # TODO: should we also check if setup would be overridden?
                         session.setup = setup
 
+                    if timeout := directives.get("timeout"):
+                        session.timeout = int(timeout)
+
+                    if timeout_expected := directives.get("timeout-expect"):
```
Comment (zupo) [inline src/tesh/extract.py:137]:
```suggestion
                    if timeout_expected := directives.get("timeout-expected"):
```

**Comment by domenkozar** [comment]
Renamed to `tesh-long-running` and implemented logic to expect only the last command to timeout. Now the output is compared even when it timeouts.

Also added `--no-debug` flag to be able to manually test this one.

Tests still fail for some reason :(

**Comment by zupo** [comment]
> Tests still fail for some reason :(

I'll try to take a look at the tests tonight.

File: src/tesh/test.py [hunk 6]
```diff
@@ -103,6 +81,45 @@ def test(filename: str, session: ShellSession, verbose: bool, debug: bool) -> No
...
+
+
+def compare_outputs(shell: pexpect.spawn, block: Block, debug: bool) -> None:
+    """Compare expected and the actual output and fail if they don't match."""
+    actual_output = get_actual_output(shell)
+    expected_output = get_expected_output(block)
+
+    if not fnmatch.fnmatch(actual_output, expected_output):
+        print("❌ Failed")  # noqa: ENC100
+        print()
```
Comment (zupo) [inline src/tesh/test.py:109]: Forgot to print failing command here.

**Review by zupo — CHANGES_REQUESTED** [review]
I pushed addition of timeout and long-running examples to README. Further fixes are needed though:

1. fail output no longer contains the command that was executed (hence `test_fail` test fails)
2. something changed in `tesh-ps1` that makes the `prompt` example in README.md to fail, and `test_prompt` as well
3. by default, when `tesh-long-running` is set, I would set `tesh-timeout` to 1, to make the UX a bit nicer (not have to wait 30 seconds)

**Comment by domenkozar** [comment]
> I pushed addition of timeout and long-running examples to README. Further fixes are needed though:
> 
> 1. fail output no longer contains the command that was executed (hence `test_fail` test fails)
> 2. something changed in `tesh-ps1` that makes the `prompt` example in README.md to fail, and `test_prompt` as well
> 3. by default, when `tesh-long-running` is set, I would set `tesh-timeout` to 1, to make the UX a bit nicer (not have to wait 30 seconds)

3. That would be surprising behavior, what if your command takes 10s to run? I think whoever is running those tests can set the timeout to lower if they get annoyed by the long running times.

**Comment by zupo** [comment]
@domenkozar: what's the status on this one? How can I help getting it ready for merge?

**Comment by domenkozar** [comment]
@blueracer should award me a superman hero badge here

**Comment by zupo** [comment]
> @blueracer-io should award me a superman hero badge here

You don't get an award for removing the 30sec slowdown you introduced in a recent PR, lol :P

**Comment by zupo** [comment]
Fail for me locally with:

```console
➜  tesh git:(88385da) tesh README.md 
📄 Checking README.md
Traceback (most recent call last):
  File "/nix/store/mvlayd2cp4lpgwfsmbdhva7ssa3h9vwn-python3.10-poetry2nix-env-scripts/bin/.tesh-wrapped", line 13, in <module>
    sys.exit(tesh())
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/zupo/work/tesh/src/tesh/__init__.py", line 46, in tesh
    sessions = extract(f)
  File "/Users/zupo/work/tesh/src/tesh/extract.py", line 80, in extract
    raise RuntimeError("Hit end-of-file prematurely. Syntax error?")
RuntimeError: Hit end-of-file prematurely. Syntax error?
```

**Comment by zupo** [comment]
This is the culprit: https://github.com/OceanSprint/tesh/pull/30/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L205

**Comment by zupo** [comment]
@domenkozar: OK, got it to work in CI and my MacBook. Can you try if it work on your local pls?

**Comment by domenkozar** [comment]
Works for me

**Pull request merged**